docs: :memo: decision post on why to use pytest (#102)
seedcase-project/decisions: why-pytest/index.qmd

---
title: "Why Pytest"
description: |
  Testing software is a requirement for ensuring it works as intended. 
  This post explains why we choose to use pytest over the built-in 
  unittest framework for testing Python code.
date: "2024-07-17"
categories:
- test
- develop
---

::: content-hidden
Use other decision posts as inspiration to writing these. Leave the
content-hidden sections in the text for future reference.
:::

## Context and problem statement

::: content-hidden
State the context and some background on the issue, then write a
statement in the form of a question for the problem.
:::

Testing software is a basic requirement for assessing its functionality,
accuracy, and reliability. There are many categories of [software

> testing](https://en.wikipedia.org/wiki/Software_testing) that include

Reviewer: ```suggestion
testing](https://en.wikipedia.org/wiki/Software_testing), including
```

Reviewer: For me the original is a bit misleading somehow.

integration testing, unit testing, or system testing. This post is
focused on unit testing.

Since one of our [development
principles](https://design.seedcase-project.org/style/development) is to

> use a Test-Driven Development workflow, we need a testing framework when

Reviewer 2: ```suggestion
use a test-driven development workflow, we need a testing framework when
```
Right?

::: content-hidden
List some reasons for why we need to make this decision and what things
have arisen that impact work.
:::

-   Knowing that software works as intended requires some form of
    testing.
-   Developing software is easier when we can compare the expected
    output with the actual output in a formal testing approach.
-   Identifying bugs is easier when there are tests that check the
    expected functionality of our code.
-   Identifying breaking changes is also easier when there are tests.

## Considered options

::: content-hidden
List and describe some of the options, as well as some of the benefits
and drawbacks for each option.
:::

Django's default unit testing framework is
[unittest](https://docs.python.org/3/library/unittest.html), which is
also Python's built-in testing framework. Aside from the default
unittest, there is really only one other alternative for unit testing:

-   [pytest](https://docs.pytest.org/en/8.2.x/) with
    [django-pytest](https://pytest-django.readthedocs.io/en/latest/)

### unittest

[unittest](https://docs.python.org/3/library/unittest.html) is the
built-in framework in Python. It was inspired by JUnit (the unit testing

> framework in Java) and has a similar structure to it, for better or
> worse.

Reviewer 2: ```suggestion
framework in Java) and has a similar structure to it.
```

::::: columns
::: column
#### Benefits

-   Comes by default with Django and Python.
-   Documentation on the official Django website uses unittest.
:::

::: column
#### Drawbacks

> -   Requires more extra code to setup and structure tests. For instance,

Reviewer: ```suggestion
-   Requires more extra code to set up and structure tests. For instance,
```

    all tests need to be a `class` object and some testing needs
    `setUp()` and `tearDown()` methods.
-   The test output is not very readable and beginner friendly.

> -   Creating tests takes more effort and time.

Reviewer 2: more effort and time as compared to *what*? :)

> -   Creating tests takes more effort and time.

Author: ```suggestion
-   Creating tests takes more effort and time compared to alternatives.
```

> -   Writing tests follows more of the object-oriented programming style,

Reviewer: ```suggestion
-   Writing tests follows more of an object-oriented programming style,
```

Reviewer: For me at least.

>     which can be harder to reason about and design tests for.

Author: This is what I've read about across multiple sources, the OOP in general is harder to test and that writing things following an OOP style is itself harder to design. E.g. https://softwareengineering.stackexchange.com/questions/351815/why-is-unit-testing-harder-in-object-oriented-programming-compared-to-functional and https://www.geeksforgeeks.org/why-is-unit-testing-harder-in-oop/

>     which can be harder to reason about and design tests for.

Author: ```suggestion
    which can be harder to reason about and design tests for 
    (for multiple, highly technical reasons outside the scope of
    this post)
```

>     which can be harder to reason about and design tests for.

Reviewer 2: Great, thanks!

:::
:::::

### pytest

[pytest](https://docs.pytest.org/en/8.2.x/), along with the [Django
plugin](https://pytest-django.readthedocs.io/en/latest/), is a newer
testing framework that was built to improve on the limitations of the
default unittest framework.

::::: columns
::: column
#### Benefits

-   Integrates with [unittest style
    tests](https://docs.pytest.org/en/7.1.x/how-to/unittest.html), so
    don't need to refactor all existing tests.
-   Has easier to read output of the test results.
-   Uses a simpler syntax for writing tests, with a simple `assert`
    statement.
-   Created and developed more recently and from scratch, so is better

>     designed than unittest.

Author: ```suggestion
    designed than unittest for Python (since unittest was modeled off of Java's framework).
```

-   While it isn't the default unit testing framework in Python, it is
    widely used.
:::

::: column
#### Drawbacks

-   Requires installing two additional packages: `pytest` and
    `pytest-django`.
-   Tests written in the pytest style can't be used by other non-unit

>     testing frameworks.

Reviewer: I guess it's not 100% obvious to me what this means 😅 . I would understand "other unit testing frameworks"

> -   Tests written in the pytest style can't be used by other non-unit
>     testing frameworks.

Author: ```suggestion
-   Integrating pytest with other types of testing frameworks can be
		a bit difficult and may require some configuration.
```
